src/aspire/classification/rir_class2d.py

```diff
@@ -0,0 +1,577 @@
+import logging
+
+import matplotlib.pyplot as plt
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+from tqdm import tqdm
+
+from aspire.basis import FSPCABasis
+from aspire.classification import Class2D
+from aspire.classification.legacy_implementations import (
+    bispec_2drot_large,
+    pca_y,
+    rot_align,
+)
+from aspire.image import Image
+from aspire.numeric import ComplexPCA
+from aspire.operators import BlkDiagMatrix
+from aspire.utils.random import rand
+
+logger = logging.getLogger(__name__)
+
+
+class RIRClass2D(Class2D):
+    def __init__(
+        self,
+        src,
+        pca_basis,
+        fspca_components=400,
+        alpha=1 / 3,
+        sample_n=4000,
+        bispectrum_componenents=300,
+        n_nbor=100,
+        n_classes=50,
+        bispectrum_freq_cutoff=None,
+        large_pca_implementation="legacy",
+        nn_implementation="legacy",
+        bispectrum_implementation="legacy",
+        dtype=None,
+        seed=None,
+    ):
+        """
+        Constructor of an object for classifying 2D images using
+        Rotationally Invariant Representation (RIR) algorithm.
+
+        At a high level this consumes a Source instance `src`,
+        and a FSPCA Basis `pca_basis`.
+
+        Yield class averages by first performing `classify`,
+        then performing `output`.
+
+        Z. Zhao, Y. Shkolnisky, A. Singer, Rotationally Invariant Image Representation
+        for Viewing Direction Classification in Cryo-EM. (2014)
+
+        :param src: Source instance
+        :param pca_basis: FSPCA Basis instance
+        :param fspca_components: Components (top eigvals) to keep from full FSCPA, default truncates to  400.
+        :param alpha: Amplitude Power Scale, default 1/3 (eq 20 from  RIIR paper).
+        :param sample_n: A number and associated method used to confuse your enemies.
```

Comment:
:stuck_out_tongue:

Reply:
added a more appropriate comment.  Pragmatically, if this doesn't offer a lot of value in practice we might just remove it... but I think we need to actually push a lot of real data through to know that...